cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/lineage/LineageOperationsProcessor.java

```diff
@@ -129,34 +150,70 @@ public LineageOperationsProcessor(Set<Connection> stageConnections,
     return processedOperations;
   }
 
-  private void addMergeOperation(Set<String> stageInputs,
-                                 Map<String, Operation> processedOperations) {
+  /**
+   * Create the implicit merge operations. Each merge operation will have a prefix with
+   * {stage1-name},(stage2-name),{other-stage-name}.merge, appended with the field name.
+   * Each merge operation can be seen as an identity transform for all the fields appeared in the outputs in the
+   * parent stages.
+   * For example, if a pipeline looks like this:
+   * src1 -> t1(a,b,c) --|
+   *                     |-> t3(not joiner) -> sink
+   * src2 -> t2(a,b,c) --|
+   * At stage transform3, 3 merge operations are generated for each field a,b,c:
+   * 1. name: t1,t2.merge.a, input fields: t1.a, t2.a, output fields: a
+   * 2. name: t1,t2.merge.b, input fields: t1.b, t2.b, output fields: b
+   * 3. name: t1,t2.merge.c, input fields: t1.c, t2.c, output fields: c
+   *
+   * @param stageInputs the stage inputs, the size of this set be greater than 1
+   * @param processedOperations the processed operations that collect all the result
+   */
+  private void addMergeOperation(Set<String> stageInputs, Map<String, Operation> processedOperations) {
     Set<String> sortedInputs = new TreeSet<>(stageInputs);
-    String mergeOperationName = prefixedOperationName(Joiner.on(SEPARATOR).join(sortedInputs), "merge");
+    String mergeOperationNamePrefix = getMergeOperationNamePrefix(sortedInputs);
     String mergeDescription = "Merged stages: " + Joiner.on(",").join(sortedInputs);
-    if (processedOperations.containsKey(mergeOperationName)) {
-      // it is possible that same stages act as an input to multiple stages.
-      // we should still only add single merge operation for them
-      return;
-    }
-    List<InputField> inputFields = new ArrayList<>();
+
+    Map<String, List<InputField>> fieldNameMap = new LinkedHashMap<>();
```

Comment:
can you add some comments describing what the key and values are. An example would help as well.

Reply:
Added comments